### DISCUSSION TITLE
SQL Statement

### DISCUSSION THREAD

**Discussion by lupe120** [lf-edge/ekuiper#1980] [Q&A]
Hello,

Is it possible for this SQL statement:

(1)
 select * from demo GROUP BY COUNTWINDOW(14,1)  HAVING  COLLECT(speed)[3]- COLLECT(speed)[2] !=0

To only capture the first event triggered?

-----------------------------------------------------------------------------------------------------------------------------

Referring to SQL statement below:

(2)
select * from demo GROUP BY COUNTWINDOW(21,1)  HAVING  COLLECT(speed)[10]- COLLECT(speed)[9] !=0

This sql statement will capture data  if like the stream is as below:

![image](https://github.com/lf-edge/ekuiper/assets/116304794/c384ba60-7ec6-4c6e-bca5-a40451326ed7)

The yellow will be captured as there is a speed change in event 1

Because of the size of the countwindow, the following below won't be captured. The speed change happens at event 3(too early)

![image](https://github.com/lf-edge/ekuiper/assets/116304794/60e869df-3198-470b-8756-df023fc5a143)

Is it possible for this SQL statement to be modified to also catch the scenario where there is a speed change, but it occurs too early and there is not enough data points for it to work with 'GROUP BY COUNTWINDOW(21,1)' 

One solution I had is to use additional rules and sql statements, like the first one I had mentioned. 
With this SQL statement I could use to help(would also need other sql statements:

(1)
 select * from demo GROUP BY COUNTWINDOW(14,1)  HAVING  COLLECT(speed)[3]- COLLECT(speed)[2] !=0

The issue for me, is that it will continue to capture events. I would only need it to capture the first. Then it should stop functioning. 

Second solution Another would be if I could combine, one or more sql statements in the same rule. 
Could I combine the following:

(1)
 select * from demo GROUP BY COUNTWINDOW(14,1)  HAVING  COLLECT(speed)[3]- COLLECT(speed)[2] !=0 (with this sql only capturing the first event that is triggered)

And 
(2)
select * from demo GROUP BY COUNTWINDOW(21,1)  HAVING  COLLECT(speed)[10]- COLLECT(speed)[9] !=0

Therefore, I would be hoping I could capture data points from a stream as such below

![image](https://github.com/lf-edge/ekuiper/assets/116304794/e179cb84-ea16-4783-a21e-23cafc0b0b73)

Yellow highlight is what would get captured, first section captured would come from (1) SQL statement
And next section captured would come from (2) SQL statement.

**Comment by Rui-Gan** [comment]
Based on my understanding of your request, there are two main criteria.  First, for each window of 21 data points, if the speed value in the 11th data point is different from that in the 10th data point, the entire window of 21 data points needs to be outputted.  Second, when the total number of data points is less than 21, for each window of 14 data points, if the speed value in the fourth data point is different from that in the third data point, the entire window of 14 data points needs to be outputted.
Is my understanding above correct？

**Reply by lupe120** [reply]
Yes, that is correct.

**Comment by Rui-Gan** [comment]
For the first solution above, which is to use additional rules, I think your sql statement could be written like this:
```
select * from demo GROUP BY COUNTWINDOW(14,1) HAVING COLLECT(speed)[3]- COLLECT(speed)[2] !=0 AND count(lag(*)) <= 0;
```

**Reply by lupe120** [reply]
This statement works except, I only would like for it to capture the first speed change and ignore the rest or stop rule completely. The statment you provided will continue captruing windows that meet the criteria.

My solution is to add a filter such as 

select * from demo GROUP BY COUNTWINDOW(14,1) FILTER(where event < 14) HAVING COLLECT(speed)[3]- COLLECT(speed)[2] !=0 AND count(lag(*)) <= 0;

As now we would ignore events 14 and on. Only capturing the first speed change.

Can this statement be done in another way, perhaps more efficient. My only concern is that i will still have a rule actively running that will never capture anything.

Is it possible to specify a rule stop within a rule's sql statement?

-------------------------------------------------------------------------------------------------------------------------------------

Another follow up question I have is regarding this sql statement:

select * from demo GROUP BY COUNTWINDOW(21,1) HAVING COLLECT(speed)[10]- COLLECT(speed)[9] !=0

This will capture as such: 
(with the highlighted events being the first window captured)

![image](https://github.com/lf-edge/ekuiper/assets/116304794/e4b29f96-7361-4dcf-ae2d-6f8e9e8f386c)


in the snippet, there is a speed change at 32 seconds which will capture

10 events before the speed change, speed change, and 10 events after as intended(a total of 21)

And we can also see that at 35 seconds there is another speed change
![image](https://github.com/lf-edge/ekuiper/assets/116304794/5caa62ce-a4aa-4460-a856-390ba11c6ef7)

thus another window to be captured, starting from the (duplicate event):

{"speed":66,"speed_diff":0,"time":25.038909435272217} till the end
![image](https://github.com/lf-edge/ekuiper/assets/116304794/dcb18268-2756-43df-b019-7de103ce8c56)


because this speed change occurs within the first speed change, the 2nd window outputted has duplicate events with the 1st window captured. We can tell by looking at the events I have highlighted from 2nd window captured. We can observe that the time stamps are the exact same time stamps from the first window. Is it currently possible to ignore/avoid these duplicate events from being displayed/published. 

![image](https://github.com/lf-edge/ekuiper/assets/116304794/3d495c08-a10a-47ba-84b9-57be5e57fd08)


Is there any way to avoid these duplicate events if the previous window already captured them? Like an overlapping capability?

I am not seeing any windowing option that can help, although it is somewhat stated here, https://ekuiper.org/docs/en/latest/sqls/windows.html#count-window, but I am not having any success.

**Reply by Rui-Gan** [reply]
> Is it possible to specify a rule stop within a rule's sql statement?

In my view, eKuiper does not currently offer support for specifying a rule stop within a rule's SQL statement.

> Is there any way to avoid these duplicate events if the previous window already captured them? Like an overlapping capability?

Currently, there is no clear solution to avoid capturing duplicate events in overlapping windows.   However, there are some potential approaches that may help address the issue.

One option is to use a Rule Pipeline to break down the complex problem into stages and apply different rules to process the data at each stage. 

Another option is to use  `latest(ts) on (when changed)` to filter out events within the window that are too close in time to the current event.
